Desugared lambda expressions involving unlifted types won't typecheck

This `th-desugar` example involving unlifted types typechecks:

```hs
{-# LANGUAGE MagicHash #-}
{-# LANGUAGE TemplateHaskell #-}
{-# OPTIONS_GHC -ddump-splices #-}
module Foo where

import GHC.Exts (Int#)
import Language.Haskell.TH.Desugar

$(do decs <- [d| f :: Int# -> Int# -> ()
                 f 27# 42# = ()
               |]
     ddecs <- dsDecs decs
     pure (sweeten ddecs))
```

However, if I change `f` to use a lambda expression:

```hs
                 f = \27# 42# -> ()
```

Then the generated code no longer typechecks:

```
[1 of 1] Compiling Foo              ( Foo.hs, interpreted )
Foo.hs:(9,2)-(13,26): Splicing declarations
    do decs_a3w1 <- [d| f_a3w0 :: Int# -> Int# -> ()
                        f_a3w0 = \ 27# 42# -> () |]
       ddecs_a3w2 <- dsDecs decs_a3w1
       pure (sweeten ddecs_a3w2)
  ======>
    f_a3xc :: Int# -> Int# -> ()
    f_a3xc
      = \ arg_6989586621679023383_a3xe arg_6989586621679023385_a3xg
          -> case
                 ((,) arg_6989586621679023383_a3xe) arg_6989586621679023385_a3xg
             of {
               (,) 27# 42# -> () }

Foo.hs:9:2: error:
    • Couldn't match a lifted type with an unlifted type
      When matching types
        a0 :: *
        Int# :: TYPE 'GHC.Types.IntRep
    • In the first argument of ‘(,)’, namely
        ‘arg_6989586621679023383_a3xe’
      In the expression:
        ((,) arg_6989586621679023383_a3xe) arg_6989586621679023385_a3xg
      In the expression:
        case
            ((,) arg_6989586621679023383_a3xe) arg_6989586621679023385_a3xg
        of {
          (,) 27# 42# -> () }
  |
9 | $(do decs <- [d| f :: Int# -> Int# -> ()
  |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
```

As the `-ddump-splices` output indicates, this is because we desugar `\27# 42# -> ()` to `\arg1 arg2 -> case (arg1, arg2) of (27#, 42#) -> ()`. This uses a lifted pair `(,)`, however, which won't work for unlifted types like `Int#`.

I can think of a couple of solutions to this problem:

1. Make the desugaring use an unlifted pair `(#,#)`. This would require desugared code to enable `UnboxedTuples` in surprising places, however. Moreover, the desugared code would not work in old versions of GHCi without enabling `-fobject-code`.
2. Instead of using a tuple, instead `let`-bind an auxiliary function which performs the match:

   ```hs
   f :: Int# -> Int# -> ()
   f = \arg1 arg2 ->
         let aux 27# 42# = ()
          in aux arg1 arg2
   ```

   This avoids the use of tuples altogether. I'm not yet clear if this would have ramifications for `singletons`, however.

Upon further thought, solution (2) won't work in the presence of GADTs. Given this data type:

```hs
data T a where
  MkT :: Int -> T Int
```

This will typecheck:

```hs
f :: T a -> T a -> a
f = \(MkT x1) (MkT x2) -> x1 + x2
```

But this won't:

```hs
f :: T a -> T a -> a
f = \t1 t2 ->
      let aux (MkT x1) (MkT x2) = x1 + x2
      in aux t1 t2
```
```
Bug.hs:10:25: error:
    • Couldn't match expected type ‘p’ with actual type ‘T a0’
        ‘p’ is untouchable
          inside the constraints: a1 ~ Int
          bound by a pattern with constructor: MkT :: Int -> T Int,
                   in an equation for ‘aux’
          at Bug.hs:10:16-21
      ‘p’ is a rigid type variable bound by
        the inferred type of aux :: T a1 -> p -> p1
        at Bug.hs:10:11-41
      Possible fix: add a type signature for ‘aux’
    • In the pattern: MkT x2
      In an equation for ‘aux’: aux (MkT x1) (MkT x2) = x1 + x2
      In the expression: let aux (MkT x1) (MkT x2) = x1 + x2 in aux t1 t2
    • Relevant bindings include
        aux :: T a1 -> p -> p1 (bound at Bug.hs:10:11)
   |
10 |       let aux (MkT x1) (MkT x2) = x1 + x2
   |                         ^^^^^^

Bug.hs:10:35: error:
    • Couldn't match expected type ‘p1’ with actual type ‘Int’
        ‘p1’ is untouchable
          inside the constraints: a0 ~ Int
          bound by a pattern with constructor: MkT :: Int -> T Int,
                   in an equation for ‘aux’
          at Bug.hs:10:25-30
      ‘p1’ is a rigid type variable bound by
        the inferred type of aux :: T a1 -> p -> p1
        at Bug.hs:10:11-41
      Possible fix: add a type signature for ‘aux’
    • In the expression: x1 + x2
      In an equation for ‘aux’: aux (MkT x1) (MkT x2) = x1 + x2
      In the expression: let aux (MkT x1) (MkT x2) = x1 + x2 in aux t1 t2
    • Relevant bindings include
        aux :: T a1 -> p -> p1 (bound at Bug.hs:10:11)
   |
10 |       let aux (MkT x1) (MkT x2) = x1 + x2
   |                                   ^^^^^^^
```

> Contrary to your assumption, using `UnboxedTupE` does not require `-XUnboxedTuples`.

Doesn't it? If I try this:

```hs
{-# LANGUAGE TemplateHaskell #-}
module Foo where

import Language.Haskell.TH

f :: ()
f = let x = $(pure (UnboxedTupE [Just (LitE (CharL 'a')), Just (LitE (CharL 'b'))])) in ()
```

Then GHC rejects it without `-XUnboxedTuples`:

```
$ ghci-9.2 Foo.hs
GHCi, version 9.2.2: https://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /home/rscott/.ghci
[1 of 1] Compiling Foo              ( Foo.hs, interpreted )

Foo.hs:7:9: error:
    • Illegal unboxed tuple type as function argument: (# Char, Char #)
      Perhaps you intended to use UnboxedTuples
    • When checking the inferred type
        x :: (# Char, Char #)
      In the expression: let x = ((# 'a', 'b' #)) in ()
      In an equation for ‘f’: f = let x = ((# 'a', 'b' #)) in ()
  |
7 | f = let x = $(pure (UnboxedTupE [Just (LitE (CharL 'a')), Just (LitE (CharL 'b'))])) in ()
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

> One remaining problem, though, is that you evidently need `TupleSections` to use e.g. `UnboxedTupE [Nothing, Nothing]`, which is a shame.

This is true, but I don't think solution (1) would ever generate code involving tuple sections. All of the uses of `UnboxedTupE` would be fully saturated, unless I'm overlooking something.

It turns out that there actually _is_ an example of desugared code that would require the use of the `UnboxedTuples` extension, and it involves match flattening. In this example:

```hs
{-# LANGUAGE TemplateHaskell #-}
{-# OPTIONS_GHC -ddump-splices #-}
module Bug where

import Language.Haskell.TH
import Language.Haskell.TH.Desugar

data Pair a b = MkPair a b

$(pure [])

g :: String
g =
  $(do e1 <- [| let f ~(MkPair x y) = x ++ y
                in f (MkPair "a" "b")
             |]
       e2 <- dsExp e1
       e3 <- scExp e2
       pure (sweeten e3))
```

The desugared code is:

```
..\Bug.hs:(14,5)-(19,24): Splicing expression
    do e1_aCde <- [| let
                       f_aCdb ~(MkPair x_aCdc y_aCdd) = x_aCdc ++ y_aCdd
                     in f_aCdb (MkPair "a" "b") |]
       e2_aCdf <- dsExp e1_aCde
       e3_aCdg <- scExp e2_aCdf
       pure (sweeten e3_aCdg)
  ======>
    let
      f_aCeK _arg_6989586621679156781_aCeO
        = let
            tuple_6989586621679156789_aCeW
              = case _arg_6989586621679156781_aCeO of {
                  MkPair _x_6989586621679156783_aCeQ _y_6989586621679156785_aCeS
                    -> let x_aCeL = _x_6989586621679156783_aCeQ in
                       let y_aCeM = _y_6989586621679156785_aCeS in ((#,#) x_aCeL) y_aCeM }
            x_aCeL
              = case tuple_6989586621679156789_aCeW of {
                  (#,#) proj_6989586621679156791_aCeY _
                    -> proj_6989586621679156791_aCeY }
            y_aCeM
              = case tuple_6989586621679156789_aCeW of {
                  (#,#) _ proj_6989586621679156793_aCf0
                    -> proj_6989586621679156793_aCf0 }
          in ((++) x_aCeL) y_aCeM
    in f_aCeK ((MkPair "a") "b")
```

Notice that `tuple_6989586621679156789_aCeW` returns an unboxed tuple. As a result, GHC infers an unboxed tuple type for it, and it complains as such:

```
..\Bug.hs:14:5: error:
    • Illegal unboxed tuple type as function argument: (# a0, b0 #)
      Perhaps you intended to use UnboxedTuples
    • When checking the inferred type
        tuple_6989586621679156789_aCeW :: (# a0, b0 #)
      In the expression:
        let
          tuple_6989586621679156789_aCeW
            = case _arg_6989586621679156781_aCeO of {
                MkPair _x_6989586621679156783_aCeQ _y_6989586621679156785_aCeS
                  -> ... }
          x_aCeL
            = case tuple_6989586621679156789_aCeW of {
                (#,#) proj_6989586621679156791_aCeY _ -> ... }
          y_aCeM
            = case tuple_6989586621679156789_aCeW of {
                (#,#) _ proj_6989586621679156793_aCf0 -> ... }
        in ((++) x_aCeL) y_aCeM
      In an equation for ‘f_aCeK’:
          f_aCeK _arg_6989586621679156781_aCeO
            = let
                tuple_6989586621679156789_aCeW = ...
                x_aCeL = ...
                ....
              in ((++) x_aCeL) y_aCeM
   |
14 |   $(do e1 <- [| let f ~(MkPair x y) = x ++ y
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
```

The relevant code is in [`mkSelectorDecs`](https://github.com/goldfirere/th-desugar/blob/8357a4f40aa6ec380b321de6a4a675779041fa9f/Language/Haskell/TH/Desugar/Match.hs#L180-L228). Interestingly, the comments for that function claim that the code is similar to what GHC itself does during compilation. I wonder how it handles this example?

Ah, I figured out the answer to my own question. Here is a variant of that example which I cooked up to determine how `mkSelectorDecs` behaves for unlifted types:

```hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE KindSignatures #-}
{-# OPTIONS_GHC -ddump-simpl #-}
module Bug where

import GHC.Exts

data Pair (a :: TYPE UnliftedRep) (b :: TYPE UnliftedRep) = MkPair a b

f :: Pair a b -> Pair b a
f ~(MkPair x y) = MkPair y x
```

The answer is surprisingly straightforward: it just chooses not to even try:

```
$ ghc Bug.hs
[1 of 1] Compiling Bug              ( Bug.hs, Bug.o )

Bug.hs:11:4: error:
    A lazy (~) pattern cannot bind variables of unlifted type.
    Unlifted variables:
      x :: a
      y :: b
   |
11 | f ~(MkPair x y) = MkPair y x
   |    ^^^^^^^^^^^^
```

In that case, I'll apply the same restriction to `th-desugar`'s implementation of `mkSelectorDecs` as well.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -5,6 +5,10 @@ Version 1.14 [????.??.??]
 -------------------------
 * Drop support for GHC 7.8 and 7.10. As a consequence of this, the
   `strictToBang` was removed as it no longer serves a useful purpose.
+* Desugared lambda expressions and guards that bind multiple patterns can now
+  have patterns with unlifted types. The desugared code uses `UnboxedTuples` to
+  make this possible, so if you load the desugared code into GHCi on prior to
+  GHC 9.2, you will need to enable `-fobject-code`.
 * Fix an inconsistency which caused non-exhaustive `case` expressions to be
   desugared into uses of `EmptyCase`. Non-exhaustive `case` expressions are now
   desugared into code that throws a "`Non-exhaustive patterns in...`" error at

--- a/Language/Haskell/TH/Desugar/Core.hs
+++ b/Language/Haskell/TH/Desugar/Core.hs
@@ -298,8 +298,8 @@ mkDLamEFromDPats pats exp
   = return $ DLamE names exp
   | otherwise
   = do arg_names <- replicateM (length pats) (newUniqueName "arg")
-       let scrutinee = mkTupleDExp (map DVarE arg_names)
-           match     = DMatch (mkTupleDPat pats) exp
+       let scrutinee = mkUnboxedTupleDExp (map DVarE arg_names)
+           match     = DMatch (mkUnboxedTupleDPat pats) exp
        return $ DLamE arg_names (DCaseE scrutinee [match])
   where
     stripDVarP_maybe :: DPat -> Maybe Name
@@ -1048,7 +1048,7 @@ dsClauses n (Clause pats (NormalB exp) where_decs : rest) = do
   return $ DClause pats' exp'' : rest'
 dsClauses n clauses@(Clause outer_pats _ _ : _) = do
   arg_names <- replicateM (length outer_pats) (newUniqueName "arg")
-  let scrutinee = mkTupleDExp (map DVarE arg_names)
+  let scrutinee = mkUnboxedTupleDExp (map DVarE arg_names)
   clause <- DClause (map DVarP arg_names) <$>
               (DCaseE scrutinee <$> foldrM (clause_to_dmatch scrutinee) [] clauses)
   return [clause]
@@ -1060,7 +1060,7 @@ dsClauses n clauses@(Clause outer_pats _ _ : _) = do
       exp <- dsBody body where_decs failure_exp
       (pats', exp') <- dsPatsOverExp pats exp
       uni_pats <- fmap getAll $ concatMapM (fmap All . isUniversalPattern) pats'
-      let match = DMatch (mkTupleDPat pats') exp'
+      let match = DMatch (mkUnboxedTupleDPat pats') exp'
       if uni_pats
       then return [match]
       else return (match : failure_matches)
@@ -1368,20 +1368,40 @@ reorderFields' ds_thing con_name field_names_types field_things deflts =
         Nothing -> return $ deflt : rest'
     reorder (_ : _) [] = error "Internal error in th-desugar."
 
+-- mkTupleDExp, mkUnboxedTupleDExp, and friends construct tuples, avoiding the
+-- use of 1-tuples. These are used to create auxiliary tuple values when
+-- desugaring pattern-matching constructs to simpler forms.
+-- See Note [Auxiliary tuples in pattern matching].
+
 -- | Make a tuple 'DExp' from a list of 'DExp's. Avoids using a 1-tuple.
 mkTupleDExp :: [DExp] -> DExp
 mkTupleDExp [exp] = exp
 mkTupleDExp exps = foldl DAppE (DConE $ tupleDataName (length exps)) exps
+
+-- | Make an unboxed tuple 'DExp' from a list of 'DExp's. Avoids using a 1-tuple.
+mkUnboxedTupleDExp :: [DExp] -> DExp
+mkUnboxedTupleDExp [exp] = exp
+mkUnboxedTupleDExp exps = foldl DAppE (DConE $ unboxedTupleDataName (length exps)) exps
 
 -- | Make a tuple 'Exp' from a list of 'Exp's. Avoids using a 1-tuple.
 mkTupleExp :: [Exp] -> Exp
 mkTupleExp [exp] = exp
 mkTupleExp exps = foldl AppE (ConE $ tupleDataName (length exps)) exps
 
+-- | Make an unboxed tuple 'Exp' from a list of 'Exp's. Avoids using a 1-tuple.
+mkUnboxedTupleExp :: [Exp] -> Exp
+mkUnboxedTupleExp [exp] = exp
+mkUnboxedTupleExp exps = foldl AppE (ConE $ unboxedTupleDataName (length exps)) exps
+
 -- | Make a tuple 'DPat' from a list of 'DPat's. Avoids using a 1-tuple.
 mkTupleDPat :: [DPat] -> DPat
 mkTupleDPat [pat] = pat
 mkTupleDPat pats = DConP (tupleDataName (length pats)) [] pats
+
+-- | Make an unboxed tuple 'DPat' from a list of 'DPat's. Avoids using a 1-tuple.
+mkUnboxedTupleDPat :: [DPat] -> DPat
+mkUnboxedTupleDPat [pat] = pat
+mkUnboxedTupleDPat pats = DConP (unboxedTupleDataName (length pats)) [] pats
 
 -- | Is this pattern guaranteed to match?
 isUniversalPattern :: DsMonad q => DPat -> q Bool
@@ -1741,4 +1761,138 @@ For sweetening:
 * For all other cases, just straightforwardly sweeten
   `DForallT DForallInvis tvbs ty` to `ForallT tvbs [] ty` and
   `DConstrainedT ctxt ty` to `ForallT [] ctxt ty`.
+
+Note [Auxiliary tuples in pattern matching]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+th-desugar simplifies the overall treatment of pattern matching in two
+notable ways:
+
+1. Lambda expressions only bind variables and do not directly perform pattern
+   matching. For example, this:
+
+     \True False -> ()
+
+   Roughly desugars to:
+
+     \x y -> case (x, y) of
+               (True, False) -> ()
+               _             -> error "Non-exhaustive patterns"
+2. th-desugar does not have guards, as guards are desugared into pattern
+   matches. For example, this:
+
+     f x y | True <- x
+           , False <- y
+           = ()
+
+  Roughly desugars to:
+
+    f x y = case (x, y) of
+              (True, False) -> ()
+              _             -> error "Non-exhaustive patterns"
+
+In both of these examples, there are multiple expressions being matched on
+simultaneously. When desugaring these examples to `case` expressions, we need a
+construct that allows us to group these patterns together. Auxiliary tuples are
+one way to accomplish this.
+
+While this use of tuples works well when the arguments have lifted types, such
+as Bool, it doesn't work when the arguments have unlifted types, such as Int#.
+Imagine desugaring this lambda expression, for instance:
+
+  \27# 42# -> ()
+
+The approach above would desugar this to:
+
+  \x y -> case (x, y) of
+            (27#, 42#) -> ()
+            _          -> error "Non-exhaustive patterns"
+
+This will not typecheck, however, as we are using _lifted_ tuples, which
+require their arguments to have lifted types. If we want to support unlifted
+types, we need a different approach.
+
+One idea that seems tempting at first is to create an auxiliary `let`
+expression, e.g.,
+
+  \x y ->
+    let aux 27# 42# = ()
+     in aux x y
+
+This avoids having to use lifted tuples, but it creates a new problem: type
+inference. In the general case, auxiliary `let` expressions aren't enough to
+handle GADT pattern matches, such as in this example:
+
+  data T a where
+    MkT :: Int -> T Int
+
+  g :: T a -> T a -> a
+  g = \(MkT x1) (MkT x2) -> x1 + x2
+
+If you desugar `g` to use an auxiliary `let` expression:
+
+  g :: T a -> T a -> a
+  g = \t1 t2 ->
+        let aux (MkT x1) (MkT x2) = x1 + x2
+        in aux t1 t2
+
+Then it will not typecheck. To make this work, you'd need to give `aux` a type
+signature. Doing this in general is tantamount to performing type inference,
+however, which is very challenging in a Template Haskell setting.
+
+Another approach, which is what th-desugar currently uses, is to use auxiliary
+_unboxed_ tuples. This is identical to the previous tuple approach, but with
+slightly different syntax:
+
+  \x y -> case (# x, y #) of
+            (# 27#, 42# #) -> ()
+            _              -> error "Non-exhaustive patterns"
+
+Unboxed tuples can handle lifted and unlifted arguments alike, so it is capable
+of handling all the examples above.
+
+You might worry that this approach would require clients of th-desugar to
+enable the UnboxedTuples extension in non-obvious places, but fortunately, this
+is not the case. For one thing, all unboxed tuples produced by th-desugar would
+be TH-generated, so we would bypass the need to enable UnboxedTuples to lex
+unboxed tuple syntax. GHC's typechecker also imposes a requirement that
+UnboxedTuples be enabled if a variable has an unboxed tuple type, but this
+never happens in th-desugar by construction. It's possible that a future
+version of GHC might be stricter about this, but it seems unlikely.
+
+There are a couple of exceptions to the general rule that auxiliary binders
+should be unboxed:
+
+1. ParallelListComp is desugared using the `mzip` function, which returns a
+   lifted pair. As a result, the variables bound in a parallel list
+   comprehension must be lifted. This is a restriction which is inherited from
+   GHC itself—https://gitlab.haskell.org/ghc/ghc/-/merge_requests/7270.
+
+2. Match flattening desugars lazy patterns that bind multiple variables to code
+   that extracts fields from tuples. For instance, this:
+
+     data Pair a b = MkPair a b
+
+     f :: Pair a b -> Pair b a
+     f ~(MkPair x y) = MkPair y x
+
+   Desugars to this (roughly) when match-flattened:
+
+     f :: Pair a b -> Pair b a
+     f p =
+       let tuple = case p of
+                     MkPair x y -> (x, y)
+
+           x = case tuple of
+                 (x, _) -> x
+
+           y = case tuple of
+                 (_, y) -> x
+
+        in MkPair y x
+
+   One could imagine using an unboxed tuple here instead, but since the
+   intermediate `tuple` value would have an unboxed tuple this, this would
+   require users of match flattening to enable UnboxedTuples. Fortunately,
+   using unboxed tuples here isn't necessary, as GHC doesn't support binding
+   variables with unlifted types in lazy patterns anyway.
 -}

--- a/Test/T158Exp.hs
+++ b/Test/T158Exp.hs
@@ -1,0 +1,14 @@
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+-- | A regression test for #158 which ensures that lambda expressions
+-- containing patterns with unlifted types desugar as expected. We define this
+-- test in its own module, without UnboxedTuples enabled, to ensure that users
+-- do not have to enable the extension themselves.
+module T158Exp where
+
+import Language.Haskell.TH.Desugar
+
+t158 :: ()
+t158 =
+  $([| (\27# 42# -> ()) 27# 42# |] >>= dsExp >>= return . expToTH)

--- a/th-desugar.cabal
+++ b/th-desugar.cabal
@@ -85,6 +85,7 @@ test-suite spec
                       ReifyTypeCUSKs
                       ReifyTypeSigs
                       Splices
+                      T158Exp
                       T159Decs
 
   build-depends:

